### PR TITLE
os/mac: MachO.dynamically_linked_libraries filtering.

### PR DESCRIPTION
**Pull request by woodruffw**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This allows us to filter out weak linkages during audits, preventing
a false error from occurring when the dylib cannot be found.

ref Homebrew/homebrew-core#6645
cc @MikeMcQuaid, @UniqMartin

### CODE REVIEW

**Comment by woodruffw** [comment]
P.S.: I noticed that we're using the `MachO` namespace in Homebrew as well as in `ruby-macho`. This hasn't been an issue yet, but it could lead to some method clobbering in the future. Thoughts on changing the module name in *os/mac/mach.rb* to `MachOShim` or something similar?

**Comment by MikeMcQuaid** [comment]
> Thoughts on changing the module name in os/mac/mach.rb to MachOShim or something similar?

Fine with me 👍

File: Library/Homebrew/os/mac/linkage_checker.rb [hunk 1]
```diff
@@ -23,7 +23,7 @@ def check_dylibs
     @keg.find do |file|
       next if file.symlink? || file.directory?
       next unless file.dylib? || file.mach_o_executable? || file.mach_o_bundle?
-      file.dynamically_linked_libraries.each do |dylib|
+      file.dynamically_linked_libraries(except: :LC_LOAD_WEAK_DYLIB).each do |dylib|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/linkage_checker.rb]: A comment above here explaining why would be good.

File: Library/Homebrew/os/mac/mach.rb [hunk 2]
```diff
@@ -51,8 +51,10 @@ def mach_data
     end
   end
 
-  def dynamically_linked_libraries
-    macho.linked_dylibs
+  def dynamically_linked_libraries(except: [])
+    lcs = macho.dylib_load_commands.reject { |lc| [*except].include?(lc.type) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/mach.rb]: Looks like you're expecting `expect` to be an array here but it's a non-array above. I'm happy with either but would be good to be consistent. Additionally, if you're expecting an array I think `expect` is equivalent to `[*except]`, no?

Comment (woodruffw) [inline Library/Homebrew/os/mac/mach.rb]: I did this to allow someone to pass either an array of symbols to filter on or a single symbol. The `[*except]` will correctly pack either a single value or an array.

That's pretty hard to read though, so I could replace it with something like this:

```ruby
unless except.is_a? Array
  except = [except]
end
```

Comment (woodruffw) [inline Library/Homebrew/os/mac/mach.rb]: Got it!

**Comment by MikeMcQuaid** [comment]
Thanks again @woodruffw!

**Comment by MikeMcQuaid** [comment]
Apologies, merged this prematurely because I didn't realise that this was reliant on a new ruby-macho.  Would be good to have tests that catch this in future.

**Comment by woodruffw** [comment]
Yikes. This was a huge oversight on my part - we never added `MachO::FatFile#dylib_load_commands`, only the corresponding `MachOFile` method. I'll make the necessary changes to `ruby-macho`, submit a revendoring PR, and then this should be good to merge.

**Comment by woodruffw** [comment]
#1469 will fix that bug. This is a difficult case to test, as `MachOFile` and `FatFile` can't be complete clones of each other (each has semantics that don't make sense for the other), but I'll add something to `ruby-macho`'s test suite that enforces expected compatibility between the two.

**Pull request merged**